data_layout.go

```diff
@@ -0,0 +1,421 @@
+	dl.vecSizeAlign[instance.size] = instance
+}
+
+func (dl *DataLayout) AddFloatingPointSizeAlignment(instance *FloatingPointSizeAlignment) {
+	dl.floatSizeAlign[instance.size] = instance
+}
+
+func (dl *DataLayout) SetAggregateAlignment(instance *AggregateAlignment) {
+	dl.aggAlign = instance
+}
+
+func (dl *DataLayout) SetFunctionPointerAlignment(instance *FunctionPointerAlignment) {
+	dl.funcPtrAlign = instance
+}
+
+func (dl *DataLayout) SetManglingStyle(style ManglingStyle) {
+	dl.mangling = style
+}
+
+func (dl *DataLayout) SetNativeIntBitWidths(bitWidths []uint64) {
+	dl.nativeIntBitWidths = bitWidths
+}
+
+func (dl *DataLayout) GetLayoutString() string {
+	layout := ""
+	if dl.isBigEndian {
+		layout += "E"
+	} else {
+		layout += "e"
+	}
+	if dl.naturalStackAlignment != 0 {
+		layout += "-S" + fmt.Sprint(dl.naturalStackAlignment)
+	}
+	if dl.programMemoryAddressSpace != 0 {
+		layout += "-P" + fmt.Sprint(dl.programMemoryAddressSpace)
+	}
+	if dl.globalVarAddressSpace != 0 {
+		layout += "-G" + fmt.Sprint(dl.globalVarAddressSpace)
+	}
+	if dl.allocaAddressSpace != 0 {
+		layout += "-A" + fmt.Sprint(dl.allocaAddressSpace)
+	}
+	if len(dl.ptrSizeAlign) > 0 {
+		for k, v := range dl.ptrSizeAlign {
+			if k == 0 {
+				layout += "-p"
+			} else {
+				layout += "-p" + fmt.Sprint(k)
+			}
+			layout += ":" + fmt.Sprint(v.size) + ":" + fmt.Sprint(v.abiAlignment) + ":" +
+				fmt.Sprint(v.preferredAlignment) + ":" + fmt.Sprint(v.addressCalculationIndex)
+		}
+	}
+	if len(dl.intSizeAlign) > 0 {
+		for k, v := range dl.intSizeAlign {
+			layout += "-i" + fmt.Sprint(k) + ":" + fmt.Sprint(v.abiAlignment) + ":" + fmt.Sprint(v.preferredAlignment)
+		}
+	}
+	if len(dl.vecSizeAlign) > 0 {
+		for k, v := range dl.vecSizeAlign {
+			layout += "-v" + fmt.Sprint(k) + ":" + fmt.Sprint(v.abiAlignment) + ":" + fmt.Sprint(v.preferredAlignment)
+		}
+	}
+	if len(dl.floatSizeAlign) > 0 {
+		for k, v := range dl.floatSizeAlign {
+			layout += "-f" + fmt.Sprint(k) + ":" + fmt.Sprint(v.abiAlignment) + ":" + fmt.Sprint(v.preferredAlignment)
+		}
+	}
+	if dl.aggAlign != nil {
+		layout += "-a:" + fmt.Sprint(dl.aggAlign.abiAlignment) + ":" + fmt.Sprint(dl.aggAlign.preferredAlignment)
+	}
+	if dl.funcPtrAlign != nil {
+		if dl.funcPtrAlign.isIndependant {
+			layout += "-Fi" + fmt.Sprint(dl.funcPtrAlign.abiAlignment)
+		} else {
+			layout += "-Fn" + fmt.Sprint(dl.funcPtrAlign.abiAlignment)
+		}
+	}
+	layout += "-m:" + string(dl.mangling)
+	if bitWidths := len(dl.nativeIntBitWidths); bitWidths > 0 {
+		layout += "-n" + fmt.Sprint(dl.nativeIntBitWidths[0])
+		for i := 1; i < bitWidths; i++ {
+			layout += ":" + fmt.Sprint(dl.nativeIntBitWidths[i])
+		}
+	}
+	if len(dl.nonIntegralPointerTypes) > 0 {
+		layout += "-ni"
+		for _, v := range dl.nonIntegralPointerTypes {
+			layout += ":" + fmt.Sprint(v)
+		}
+	}
+	return layout
+}
+
+type PointerSizeAlignment struct { // All sizes are in bits.
+	addressSpace            uint64
```

Comment:
Same comment as above, export struct fields for consistency with other parts of the `llir` API.